toolchain/sem_ir/entity_with_params_base.h

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEM_IR_ENTITY_WITH_PARAMS_BASE_H_
+#define CARBON_TOOLCHAIN_SEM_IR_ENTITY_WITH_PARAMS_BASE_H_
+
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::SemIR {
+
+// Common entity fields.
+//
+// `EntityWithParamsBase` would be a base class of entities like `Function`,
+// except that then we couldn't use named initialization (or would need to
+// disable warnings about mixing named and unnamed initialization) due to how
+// C++ handles initialization of base structs. Instead, this is composed with a
+// `Fields` struct to provide an entity's actual struct.
+//
+// For example:
+//   struct FunctionFields {
+//     ... data members ...
+//   };
+//
+//   struct Function : public EntityWithParamsBase,
+//                     public FunctionFields, public Printable<Function> {
+//     ... methods ...
+//   };
+//
+// This achieves a few things:
+//   - Allows named initialization, such as:
+//     `{{.name_id = ...}, {.function_field = ...}}`
+//   - Makes `entity.name_id` access work.
+//   - Allows passing a `EntityWithParamsBase*` when only common fields are
+//     needed.
+//   - Does all this in a way that's vanilla C++.
+struct EntityWithParamsBase {
+  auto PrintBaseFields(llvm::raw_ostream& out) const -> void {
+    out << "name: " << name_id << ", parent_scope: " << parent_scope_id;
+  }
```

Comment:
It seems a bit weird to mention the "Base" in the name here. Maybe we could just call this `PrintFields`, shadow it in the derived class when necessary, and have `Printable<T>` call `PrintFields` if it exists:

```c++
struct EntityWithParamsBase {
  auto PrintFields(llvm::raw_ostream& out) const -> void { ... }
  // ...
}

struct Function : EntityWithParamsBase, FunctionFields, Printable<Function> {
  auto PrintFields(llvm::raw_ostream& out) const -> void {
    EntityWithParamsBase::PrintFields(out);
    out << ...function fields...
  }
};
```

But I'm happy to leave that to later given that this is already a large PR.